### DISCUSSION TITLE
How to extended DefaultRecord

### DISCUSSION THREAD

**Discussion by alexukua** [vufind-org/vufind#4068] [General]
I want to override \VuFind\RecordDriver\DefaultRecord::getXML

```
php public/index.php generate extendclass VuFind\\RecordDriver\\DefaultRecord ExtentMetadata
Could not find factory for VuFind\RecordDriver\DefaultRecord
```

But this class doesn't have Factory and isn't defined in PluginManager
How to create a plugin for override this class?

**Comment by demiankatz** [comment]
What kind of records are you using? DefaultRecord is an abstract base class that is not literally instantiated anywhere. If you extend it, you will not change any behavior, because the behavior you want to change is being used by a different subclass of DefaultRecord.

Most likely, you actually want to extend SolrDefault for your purposes… but if you’re working with non-Solr-based records, or records using a more specific driver (e.g. SolrMarc), you may need to extend something different to override the XML behavior. And if you want to override the behavior across multiple record types, you may need to extend multiple classes and use a trait to share custom behavior between them.

I hope this helps, but please let me know if you still have questions or need more help figuring out the best strategy!

**Reply by alexukua** [reply]
Thank you for your answer.
My target issue is to extend metadata for OAI-PMH.  
The current realization OAI-PMH interface doesn't have information about the dc.description . I want to override \VuFind\RecordDriver\DefaultRecord::getXML for an extended set of metadata (description... etc). 

I know another way is extending  \VuFind\OAI\Server.  However easier and the right method according to the single responsibility principle ii is to extend the DefaultRecord class. 

If is not possible I will override  \VuFind\OAI\Server. 
Do you have any idea how to override DefaultRecord?

**Reply by demiankatz** [reply]
@alexukua, every record driver in VuFind is a subclass of DefaultRecord. The PHP language does not offer a mechanism for overriding a class in the middle of an inheritance chain -- you can only extend off of the end of the chain. Thus, there are two options:

1.) Modify the DefaultRecord class directly. This is obviously not a good practice for a local customization, but if you feel your changes would be of general value, you could open a pull request for consideration in a future release.

2.) Extend all classes that are used in your VuFind instance, and put the custom logic on those classes. If this is only one class (i.e. if you only have one type of record in your index), it is an easy solution. If you have a heterogeneous index with records from multiple sources, it is less easy, but you can still achieve the effect with traits, so it is not too hard.

So for example, if you have a whole index full of MARC records, you just need to extend the VuFind\RecordDriver\SolrMarc class and put your custom getXML logic there. If you have a whole index of XML records, you would extend SolrDefault instead of SolrMarc. If you had both, you could create a CustomGetXMLTrait trait which includes your custom getXML method, extend BOTH classes, and "use" the trait in both of those custom subclasses.

Does that help/make sense?
